### ISSUE TITLE
High CPU usage

### ISSUE THREAD

**Issue by behrad** [mcollina/mosca#446]
Mosca is eating a lot of CPU (80%) on a 6core high end server, ran by pm2 in clustered mode, with 10,000 connected connections + around 2 connect/disconnects per second. (keepalives are 5min, 7min, 3min)

I believe retimer is still using many `setTimeout`s for each client connections, however implementating a single internal looper (with setImediate) which checks all registered client's keep-alive timeout callbacks, and runs timed-out ones.
@mcollina Can you check this https://github.com/Krb686/nanotimer, which I think is nearest module on npm to that?

**Comment by mcollina** [comment]
Can you reproduce with [0x](http://npm.im/0x)? It is important to know where the time is spent. How many messages are exchanged?

**Comment by behrad** [comment]
Is 0x OK to be ran under production in such a heavy load?
If so, can you help me set that up and generated helpful info/logs that you may want to see?

publishes are <= 1 per sec

**Comment by mcollina** [comment]
> Is 0x OK to be ran under production in such a heavy load?

No, you will need to setup a load testing scenario (I think you already have that).

> publishes are <= 1 per sec

1 per sec from each device? 1 per sec absolutely?

**Comment by behrad** [comment]
> No, you will need to setup a load testing scenario (I think you already have that).

real clients are mobile phones, I'd been unable to perfectly simulate their connection drops in my testbed. I'll try to create one although.



> 1 per sec from each device? 1 per sec absolutely?

absolute. around 1, 2 msg/sec from server to random clients.

**Comment by behrad** [comment]
BTW many phone devices are disconnected always in real-world this way:

```
WARN:  read ECONNRESET
err: Error: read ECONNRESET
       at exports._errnoException (util.js:870:11)
       at TCP.onread (net.js:544:26)
       at TCP.<anonymous> (/root/broker/source/node_modules/loopback/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31)
client: "CLIENT-ID"
```

**Comment by mcollina** [comment]
So, mosca should be logging a lot in this scenario, right? What is mosca doing?

**Comment by behrad** [comment]
Log level is WARN in production, so I can't tell you...
We can guess processing should be two fold:

1) all users are clean:false, subscribing to 3topics each. So each close/connect will trigger multiple redis operations (which I believe would not be a problem) and multiple Qlobber updates (which may be the problem), and this operations is dispatched to all Mosca instances (both ran in current machine by cluster AND other machine) (related to #441 )

2) Keepalive timers!

**Comment by mcollina** [comment]
Can you replicate this in a controlled environment with 100 or 1000 fake clients (maybe) and check what's happening?

**Comment by behrad** [comment]
I'm planning to...

**Comment by behrad** [comment]
finally resolved the CPU eater by duplicating our load on staging servers, and that is the hasher method of `pbkdf2-password`@https://github.com/mcollina/mosca/blob/master/lib/authorizer.js#L147-L161

Only removing hashing and just passing a `true` reduces 80% CPU usage to 15% !!!
I'm planning to plug my simple authorizer which compares simple raw passwords without hashing them, however I don't know if you are planning to switch to a lower overhead authentication @mcollina

**Comment by mcollina** [comment]
@behrad the hashing thing is _built_ to be slow, as storing passwords in the clear is really a bad security practice. If it was fast, it would be fast for an attacker to crack it. 
A much better solution would be to use certificates (like AWS IoT).

TL;DR, feel free to write your own authorizer :).

**Comment by behrad** [comment]
> A much better solution would be to use certificates (like AWS IoT).

Can you give me more insights on this? 
We can use a single certificate on devices of all customer accounts (tenants) however there should be a tenant based authorization for subscriptions & publishes

**Comment by mcollina** [comment]
Look at how AWS IoT work. This is possible to achieve, but there is
currently no code for this.
Il giorno lun 2 mag 2016 alle 16:50 Behrad <notifications@github.com> ha
scritto:

**Comment by mcollina** [comment]
Should be fixed in v2.0.0.
